Sources/SwiftKafka/RDKafka/RDKafkaConfig.swift

```diff
@@ -13,18 +13,20 @@
 //===----------------------------------------------------------------------===//
 
 import Crdkafka
+import Logging
 
 /// A collection of helper functions wrapping common `rd_kafka_conf_*` functions in Swift.
 struct RDKafkaConfig {
     typealias KafkaAcknowledgementResult = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
     /// Wraps a Swift closure inside of a class to be able to pass it to `librdkafka` as an `OpaquePointer`.
-    final class CapturedClosure {
-        typealias Closure = (KafkaAcknowledgementResult?) -> Void
-        let closure: Closure
+    final class CapturedClosures {
+        typealias DeliveryReportClosure = (KafkaAcknowledgementResult?) -> Void
+        var deliveryReportClosure: DeliveryReportClosure?
 
-        init(_ closure: @escaping Closure) {
-            self.closure = closure
-        }
+        typealias LoggingClosure = (Int32, UnsafePointer<CChar>, UnsafePointer<CChar>) -> Void
```

Comment:
It seems if use `String` then string object will be constructed (with possible allocation and copying) with no respect to logging level. I think it's best to do this only if corresponding log level is enabled.

I see two options: use pointers or pass `@autoclosure`s like `() -> String` and call it inside the log message construction autoclosure.

Reply:
While I agree that it would be nice to only alloc when we know we actually log the output, your PR is currently not doing that. You unconditionally bridge the pointer into a `String` inside the closure further down. I would prefer if we do that for now unconditionally, if we see that it is a perf problem we can always change.

Reply:
Hm, seems I do not understand something. Seems need your help to understand.

If I add next extension for `String`:
```
extension String {
    init(ccString: UnsafePointer<CChar>) {
        print("ccString init")
        self.init(cString: ccString)
    }
}
```
and use it inside the `loggingClosure` instead of String(cString:), so I would expect `ccString init` to be printed if the init is called. Then I add print to `loggingClosure` to print on every log from `librdkafka`. So closure now looks like:
```
let loggingClosure: RDKafkaConfig.CapturedClosures.LoggingClosure = { level, fac, buf in
            // Mapping according to https://en.wikipedia.org/wiki/Syslog
            print("got a log of level: \(level)")
            switch level {
            case 0 ... 2: /* Emergency, Alert, Critical */
                logger.critical(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            case 3: /* Error */
                logger.error(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            case 4: /* Warning */
                logger.warning(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            case 5: /* Notice */
                logger.notice(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            case 6: /* Informational */
                logger.info(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            default: /* Debug */
                logger.debug(Logger.Message(stringLiteral: String(ccString: buf)), source: String(ccString: fac))
            }
        }
```

Running test gives me something like:
```
got a log of level: 7
got a log of level: 7
got a log of level: 7
got a log of level: 7
got a log of level: 7
got a log of level: 7
got a log of level: 7
...
```
but if I enable debug logging then I get:
```
got a log of level: 7
ccString init
ccString init
2023-06-29T13:14:50+0300 debug kafka.test : [DESTROY] [thrd:app]: Terminating instance (destroy flags none (0x0))
got a log of level: 7
ccString init
ccString init
2023-06-29T13:14:50+0300 debug kafka.test : [DESTROY] [thrd:main]: Destroy internal
...
```
So, seem I do not understand something, but seems if I pass pointers to `loggingClosure` then String will be constructed if and only if corresponding logging level is enabled.

Sorry if the question is stupid, but what do I miss?

Reply:
Ah no you are 100% right. I misread the previous code. Since the various methods on `logger` are using `@autoclosure` this will not allocate. Fair enough. Let's keep it like it is.